### PR TITLE
🧺 [1/N] Refactor `_generate` in GRPO/RLOO: list of ints instead of tensors

### PR DESCRIPTION
**Pull request by qgallouedec**
This PR is the third of a sequence of PR (after https://github.com/huggingface/trl/pull/4111, https://github.com/huggingface/trl/pull/4113 and https://github.com/huggingface/trl/pull/4114) that aims to refactor the generation part of GRPO/RLOO to allow for easier customization and ultimately tool calling

The idea with this PR is to make `_generate` return list of ints instead of tensors. This will help a lots when implementing tool calling.

Several modifications:

1. `truncate_with_protected_tokens`: instead of operating on 2D tensors (ids and mask), it will operate on sequence ids directly:

    before
    
    ```python
    >>> ids = torch.tensor([[1, 2, 3], [4, 5, 6]])
    >>> mask = torch.tensor([[1, 1, 1], [0, 0, 1]])
    >>> truncate_with_protected_tokens(ids, mask, target_length=2, protected_tokens=[])
    tensor([[2, 3], [5, 6]]), tensor([[1, 1], [0, 1]])
    ```
    
    after
    
    ```python
    >>> ids = [[1, 2, 3], [4, 5, 6]]
    >>> [truncate_with_protected_tokens(seq, target_length=2, protected_tokens=[]) for seq in ids]
    [[2, 3], [5, 6]]
    ```

2. `_generate` now returns list of ids, instead of tensor + mask
    conversion to tensor is handle in `_generate_and_score_completions`

3. The generation part is moved to a function `_generate_single_turn`, which is called by `_generate`

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1192,14 +1192,14 @@ def _generate(self, prompts: list[str], images: Optional[list]):
                 # Broadcast the completions from the main process to all processes, ensuring each process receives its corresponding slice.
                 obj_list = [payload]
                 broadcast_object_list(obj_list, from_process=0)
-                completion_ids, all_logprobs = obj_list[0]
+                all_completion_ids, all_logprobs = obj_list[0]
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1195]: more consistent naming

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1192,14 +1192,14 @@ def _generate(self, prompts: list[str], images: Optional[list]):
...
 
                 process_slice = slice(
                     self.accelerator.process_index * len(prompts),
                     (self.accelerator.process_index + 1) * len(prompts),
                 )
-                completion_ids = completion_ids[process_slice]
-                all_logprobs = all_logprobs[process_slice]
+                completion_ids = all_completion_ids[process_slice]
+                logprobs = all_logprobs[process_slice]
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1202]: same, better naming

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -1252,7 +1252,7 @@ def _generate(self, prompts: list[str], images: Optional[list]):
                 with profiling_context(self, "vLLM.generate"):
                     all_outputs = self.llm.generate(vllm_inputs, sampling_params=sampling_params, use_tqdm=False)
 
-                completion_ids = [output.token_ids for outputs in all_outputs for output in outputs.outputs]
+                all_completion_ids = [output.token_ids for outputs in all_outputs for output in outputs.outputs]
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1255]: same, better, naming

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -1264,22 +1264,12 @@ def _generate(self, prompts: list[str], images: Optional[list]):
                     # Each rank generates all outputs — we keep only our share.
                     local_rank_in_group = torch.distributed.get_rank(group=self.tp_group)
                     tp_slice = slice(local_rank_in_group * orig_size, (local_rank_in_group + 1) * orig_size)
-                    completion_ids = completion_ids[tp_slice]
-                    all_logprobs = all_logprobs[tp_slice]
+                    completion_ids = all_completion_ids[tp_slice]
+                    logprobs = all_logprobs[tp_slice]
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1268]: same, better naming

File: trl/trainer/grpo_trainer.py [hunk 5]
```diff
@@ -1264,22 +1264,12 @@ def _generate(self, prompts: list[str], images: Optional[list]):
...
-            # Pad the completions, and concatenate them with the prompts
-            completion_ids = [torch.tensor(ids, device=device) for ids in completion_ids]
-            completion_mask = [torch.ones(len(ids), device=device, dtype=torch.long) for ids in completion_ids]
-            completion_ids = pad(completion_ids, padding_value=self.pad_token_id)
-            completion_mask = pad(completion_mask, padding_value=0)
-            sampling_per_token_logps = [
-                torch.tensor(logprobs, device=device, dtype=torch.float32) for logprobs in all_logprobs
-            ]
-            sampling_per_token_logps = pad(sampling_per_token_logps, padding_value=0.0)
-
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1285]: here we want generate to return list of ints instead of padded torch tensor; for the vllm case, it's pretty easy: just remove the tensor building part.

File: trl/trainer/grpo_trainer.py [hunk 6]
```diff
@@ -1309,13 +1299,11 @@ def _generate(self, prompts: list[str], images: Optional[list]):
...
-            completion_ids = [torch.tensor(ids, device=device) for ids in completion_ids]
-            completion_ids = pad(completion_ids, padding_value=self.pad_token_id, padding_side="right")
-            prompt_ids = [torch.tensor(ids, device=device) for ids in paged_prompt_inputs.input_ids]
-            prompt_ids = pad(prompt_ids, padding_value=self.pad_token_id, padding_side="left")
+            prompt_ids = paged_prompt_inputs.input_ids
             # Restore the original attention implementation, training mode
             self.model_wrapped.config._attn_implementation = previous_attn
-            sampling_per_token_logps = None  # not used in this case
+
+            logprobs = None  # not used in this case
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: simpler naming

File: trl/trainer/grpo_trainer.py [hunk 7]
```diff
@@ -1309,13 +1299,11 @@ def _generate(self, prompts: list[str], images: Optional[list]):
                     )
                     unwrapped_model.train()  # restore training mode, as generate_batch forces eval mode
             completion_ids = [output.generated_tokens for output in all_outputs.values()]
-            completion_ids = [torch.tensor(ids, device=device) for ids in completion_ids]
-            completion_ids = pad(completion_ids, padding_value=self.pad_token_id, padding_side="right")
-            prompt_ids = [torch.tensor(ids, device=device) for ids in paged_prompt_inputs.input_ids]
-            prompt_ids = pad(prompt_ids, padding_value=self.pad_token_id, padding_side="left")
+            prompt_ids = paged_prompt_inputs.input_ids
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1305]: same as vllm, just remove the tensor building step

File: trl/trainer/grpo_trainer.py [hunk 8]
```diff
@@ -1338,29 +1326,27 @@ def _generate(self, prompts: list[str], images: Optional[list]):
...
+            is_eos = completion_ids == self.eos_token_id
+            eos_idx = torch.full((is_eos.size(0),), is_eos.size(1), dtype=torch.long, device=device)
+            eos_idx[is_eos.any(dim=1)] = is_eos.int().argmax(dim=1)[is_eos.any(dim=1)]
+            sequence_indices = torch.arange(is_eos.size(1), device=device).expand(is_eos.size(0), -1)
+            completion_mask = (sequence_indices <= eos_idx.unsqueeze(1)).int()
+
+            prompt_ids = [p[m].tolist() for p, m in zip(prompt_ids, prompt_mask.bool())]
+            completion_ids = [c[m].tolist() for c, m in zip(completion_ids, completion_mask.bool())]
+
+            logprobs = None  # not used in this case
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1345]: special case for transformers' `generate`: returns a torch tensor, and we need to extract the list of ints from it.

File: trl/trainer/grpo_trainer.py [hunk 9]
```diff
@@ -1338,29 +1326,27 @@ def _generate(self, prompts: list[str], images: Optional[list]):
...
+        # Get completion length per sequence, used for logging
+        completion_lengths = torch.tensor([len(ids) for ids in completion_ids], device=device)
         agg_completion_lengths = self.accelerator.gather(completion_lengths)
         num_items_in_batch = agg_completion_lengths.sum()  # this is required for the DAPO loss
 
-        # If mask_truncated_completions is enabled, zero out truncated completions in completion_mask
-        if self.mask_truncated_completions:
-            truncated_completions = ~is_eos.any(dim=1)
-            completion_mask = completion_mask * (~truncated_completions).unsqueeze(1).int()
-
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1362]: overlong filtering is moved outside `_generate`. The only difference that it induces is that the logged completion length and num_tokens will take into account the filtered completions, which IMO makes more sense.

File: trl/trainer/grpo_trainer.py [hunk 10]
```diff
@@ -1404,19 +1383,32 @@ def _generate_and_score_completions(
...
+        prompt_mask = pad(prompt_mask, padding_value=0, padding_side="left")
+        completion_ids = [torch.tensor(ids, device=device) for ids in completion_ids_list]
+        completion_mask = [torch.ones_like(ids, dtype=torch.long) for ids in completion_ids]
+        completion_ids = pad(completion_ids, padding_value=self.pad_token_id, padding_side="right")
+        completion_mask = pad(completion_mask, padding_value=0, padding_side="right")
+        if sampling_per_token_logps_list is not None:
+            sampling_per_token_logps = [torch.tensor(logps, device=device) for logps in sampling_per_token_logps_list]
+            sampling_per_token_logps = pad(sampling_per_token_logps, padding_value=0.0, padding_side="right")
+        else:
+            sampling_per_token_logps = None
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: Since the returned elements are list of ints, we now need to build tensors from them

File: trl/trainer/grpo_trainer.py [hunk 11]
```diff
@@ -1404,19 +1383,32 @@ def _generate_and_score_completions(
...
+
+        # If mask_truncated_completions is enabled, zero out truncated completions in completion_mask
+        if self.mask_truncated_completions:
+            completion_mask = completion_mask * (~is_truncated).unsqueeze(1).int()
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1429]: moved from generate

File: trl/trainer/grpo_trainer.py [hunk 12]
```diff
@@ -1088,15 +1087,7 @@ def _generate(self, prompts: list[str], images: Optional[list]):
...
-        prompt_inputs = self.processing_class(
-            text=prompts_text,
-            return_tensors="pt",
-            padding=True,
-            padding_side="left",
-            add_special_tokens=False,
-            **kwargs,
-        )
-        prompt_inputs = super()._prepare_inputs(prompt_inputs)
+        prompt_inputs = self.processing_class(text=prompts_text, add_special_tokens=False, **kwargs)
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: the function must now return a list of ints, so we must remove padding

File: trl/trainer/grpo_trainer.py [hunk 13]
```diff
@@ -1369,25 +1357,18 @@ def _generate(self, prompts: list[str], images: Optional[list]):
...
             term_completion_lengths = torch.zeros(1, device=device)
         self._metrics[mode]["completions/mean_terminated_length"].append(term_completion_lengths.float().mean().item())
         self._metrics[mode]["completions/min_terminated_length"].append(term_completion_lengths.float().min().item())
         self._metrics[mode]["completions/max_terminated_length"].append(term_completion_lengths.float().max().item())
 
-        return (
-            prompt_ids,
-            completion_ids,
-            prompt_mask,
-            completion_mask,
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1389]: prompt and completion masks are later inferred from the sequence lengths

File: trl/trainer/utils.py [hunk 14]
```diff
@@ -1861,63 +1861,39 @@ def unsplit_pixel_values_by_grid(batch: dict[str, Union[torch.Tensor, list[torch
...
-    Truncate tensors to target length while preserving protected tokens.
+    Truncate list to target length while preserving protected tokens.
 
     Args:
-        ids (`torch.Tensor`):
-            Input tensor of token IDs, shape (batch_size, sequence_length).
-        mask (`torch.Tensor`):
-            Input tensor of attention masks, shape (batch_size, sequence_length).
+        sequences (`list[int]`):
+            Input sequence of token IDs.
```
Comment (albertvillanova) [inline trl/trainer/utils.py]: The `sequences` name in the docstring is not aligned with the `ids` name in the signature.

File: trl/trainer/utils.py [hunk 15]
```diff
@@ -1861,63 +1861,39 @@ def unsplit_pixel_values_by_grid(batch: dict[str, Union[torch.Tensor, list[torch
...
         target_length (`int`):
-            Desired length of the output sequences.
+            Desired length of the output sequence.
         protected_tokens (`list[int]`):
             List of token IDs that should be preserved in the output.
     """
```
Comment (albertvillanova) [inline trl/trainer/utils.py:1943]:
```suggestion

    Returns:
        `list[int]`: Truncated sequence.
        
    Raises:
        `ValueError`: If `len(protected_tokens ∩ seq) > target_length`.
    """
```

File: trl/trainer/utils.py [hunk 16]
```diff
@@ -1861,63 +1861,39 @@ def unsplit_pixel_values_by_grid(batch: dict[str, Union[torch.Tensor, list[torch
...
+    num_protected = sum(1 for t in ids if t in protected_set)
+    if num_protected > target_length:
+        raise ValueError(
+            f"target_length ({target_length}) is too small for the protected tokens ({num_protected} tokens). "
+            f"Please increase target length to at least {num_protected} or disable truncation."
+        )
+    num_non_protected_needed = target_length - num_protected
+    result = []
+
+    # Iterate backward to select rightmost non-protected tokens
```
Comment (albertvillanova) [inline trl/trainer/utils.py]: Nit:
```suggestion
    # Iterate backward to select all protected tokens and rightmost non-protected tokens
```

File: trl/trainer/utils.py (hunk 14 above)
Comment (albertvillanova) [reply trl/trainer/utils.py]: Additionally, before it accepted batch_size sequences (within the tensor) and now it accepts a single sequence (list[int]). Isn't this breaking something? Some tests should be failing because of the new behavior?

Comment (qgallouedec) [reply trl/trainer/utils.py]: > Some tests should be failing because of the new behavior?

yes, tests have been updated as well, see `TruncateWithProtectedTokensTester`

Comment (qgallouedec) [reply trl/trainer/utils.py]: > The `sequences` name in the docstring is not aligned with the `ids` name in the signature.

thanks! fixed in c570fb0b
